### PR TITLE
[FLINK-16345][table-planner-blink] Fix computed column can not refer row time attribute column

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

If a computed column refers a rowtime column (the column has been defined watermark), a query on the table will be failed with the following exception:

```
[ERROR] Could not execute SQL statement. Reason:
 java.lang.AssertionError: Conversion to relational algebra failed to preserve datatypes:
 validated type:
 RecordType(VARCHAR(2147483647) CHARACTER SET "UTF-16LE" order_id, TIME ATTRIBUTE(ROWTIME) order_time, DOUBLE amount, DOUBLE amount_kg, TIMESTAMP(3) ts) NOT NULL
 converted type:
 RecordType(VARCHAR(2147483647) CHARACTER SET "UTF-16LE" order_id, TIME ATTRIBUTE(ROWTIME) order_time, DOUBLE amount, DOUBLE amount_kg, TIME ATTRIBUTE(ROWTIME) ts) NOT NULL
 rel:
 LogicalProject(order_id=[$0], order_time=[$1], amount=[$2], amount_kg=[$3], ts=[$4])
 LogicalWatermarkAssigner(rowtime=[order_time], watermark=[$1])
 LogicalProject(order_id=[$0], order_time=[$1], amount=[$2], amount_kg=[*($2, 1000)], ts=[+($1, 28800000:INTERVAL HOUR)])
 LogicalTableScan(table=[[default_catalog, default_database, orders, source: [Kafka010TableSource(order_id, order_time, amount)]]])
```

## Brief change log

- erase time indicators before translate `CatalogSourceTable` into `RelNode`s in `CatalogSourceTable#toRel`.

## Verifying this change

- added a unit plan test to reproduce this problem and verify the changes.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by wuchong** [comment]
cc @danny0405

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/CatalogSourceTable.scala [hunk 1]
```diff
@@ -78,12 +79,16 @@ class CatalogSourceTable[T](
...
+    // erase time indicator types in the rowType
+    val erasedRowType = eraseTimeIndicator(rowType, typeFactory)
 
     val tableSource = findAndCreateTableSource(flinkContext.getTableConfig.getConfiguration)
     val tableSourceTable = new TableSourceTable[T](
       relOptSchema,
       schemaTable.getTableIdentifier,
-      rowType,
+      erasedRowType,
       statistic,
```
Comment (danny0405) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/CatalogSourceTable.scala]: Why not just not to generates the TimeIndicator type when doing the row type inference of `TableSource`, it seems a little tricky to generates first then remove.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/CatalogSourceTable.scala [hunk 2]
```diff
@@ -106,7 +111,7 @@ class CatalogSourceTable[T](
     val toRexFactory = flinkContext.getSqlExprToRexConverterFactory
 
     // 2. push computed column project
-    val fieldNames = rowType.getFieldNames.asScala
+    val fieldNames = erasedRowType.getFieldNames.asScala
```
Comment (danny0405) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/CatalogSourceTable.scala]: I think the right place to fix is the `SqlToOperationConverter#createTableSchema`, before we make the computed column type inference, the time attribute should be patched up based on the watermark definitions.

**Comment by JingsongLi** [comment]
I don't fully understand this, if we just throw time indicator information, how to support process time field? Downstream operators how to know this is a process time field?

**Comment by wuchong** [comment]
> I don't fully understand this, if we just throw time indicator information, how to support process time field? Downstream operators how to know this is a process time field?

We didn't drop the time indicator information. The RelDataType from `CatalogSourceTable` to `TableScan -> Project -> Watermark` is identical, which means it keeps the time indicators. After translation, the time indicator is attached by Watermark node.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/CatalogSourceTable.scala (hunk 2 above)
Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/CatalogSourceTable.scala]: After an offline discussion with @danny0405 , we reached an consensus that erasing time indicators in CatalogSourceTable#toRel is the only way to fix this for now. So we will keep the current implementation in PR.

**Comment by libenchao** [comment]
LGTM, +1 for merging.

**Comment by wuchong** [comment]
Thanks for the reviewing @libenchao @danny0405 , will merge this.

**Pull request merged**